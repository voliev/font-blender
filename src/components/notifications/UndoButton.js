import React from "react";
import PropTypes from "prop-types";
import "./styles/UndoButton.css";

const UndoButton = ({ onClickHandler }) => (
  <button
    className="UndoButton btn"
    onClick={onClickHandler}
    value="undo"
    title="Undo"
  >
    Undo
  </button>
);

UndoButton.propTypes = {
  onClickHandler: PropTypes.func.isRequired
};

export default UndoButton;
