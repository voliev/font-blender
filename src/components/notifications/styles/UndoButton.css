.UndoButton {
  background-color: transparent;
  font-size: var(--font-xs);
  height: 36px;
  min-width: 64px;
  padding: 0;
  color: var(--orange);
  text-transform: uppercase;
}
