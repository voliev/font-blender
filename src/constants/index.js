// *** ACTION TYPES ***

// Background
export const SET_BACKGROUND_COLOR = "SET_BACKGROUND_COLOR";

// Filteres
export const SET_CATEGORIES_FILTER = "SET_CATEGORIES_FILTER";
export const SET_SORTING_ORDER = "SET_SORTING_ORDER";
export const SET_SORTING_PARAM = "SET_SORTING_PARAM";
export const SET_TITLE_FILTER = "SET_TITLE_FILTER";

// Fonts
export const FETCH_FONTS_DATA = "FETCH_FONTS_DATA";
export const FETCH_FONTS_DATA_SUCCESS = "FETCH_FONTS_DATA_SUCCESS";
export const FETCH_FONTS_DATA_FAILURE = "FETCH_FONTS_DATA_FAILURE";
export const SET_FONTS = "SET_FONTS";
export const UPDATE_FONT = "UPDATE_FONT";

// Notifications
export const ADD_NOTIFICATION = "ADD_NOTIFICATION";
export const REMOVE_NOTIFICATION = "REMOVE_NOTIFICATION";

// Showcase
export const DISPLAY_SHOWCASE_TEXT = "DISPLAY_SHOWCASE_TEXT";
export const DISPLAY_STYLES = "DISPLAY_STYLES";
export const DISPLAY_SHOWCASE_FONT_PREVIEW = "DISPLAY_SHOWCASE_FONT_PREVIEW";

// Styles
export const SET_DEFAULT_STATE = "SET_DEFAULT_STATE";
export const UNDO_RESET = "UNDO_RESET";

// Text styles
export const SET_STYLES_INHERITANCE = "SET_STYLES_INHERITANCE";
export const UPDATE_ELEMENT_STYLES = "UPDATE_ELEMENT_STYLES";

// Selected fonts
export const ADD_SELECTED_FONT = "ADD_SELECTED_FONT";
export const CLEAR_SELECTED_FONTS = "CLEAR_SELECTED_FONTS";
export const REMOVE_FONT_FROM_SELECTED = "REMOVE_FONT_FROM_SELECTED";
export const UNDO_DESELECTED = "UNDO_DESELECTED";

// Type scale
export const SET_DEFAULT_TYPE_SCALE_STATE = "SET_DEFAULT_TYPE_SCALE_STATE";
export const SET_FONT_SIZE = "SET_FONT_SIZE";
export const SET_TYPE_SCALE = "SET_TYPE_SCALE";

// *** MISCELLANEOUS ***

// Google Fonts API endpoint URL
export const GOOGLE_FONTS_API_ENDPOINT =
  "https://www.googleapis.com/webfonts/v1/webfonts";
// Fonts' bundle size that should be prefetched
export const BUNDLE_SIZE = 50;
// Available text elements for styling
export const ELEMENTS = ["body", "h1", "h2", "h3", "h4", "li", "small"];
// Font's character range that needs to be prefetched
export const CHAR_RANGE =
  "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
// Font showcase string
export const SHOWCASE_STRING =
  "That's the beauty of music. They can't get that from you...";
// Default showcase font size
export const DEFAULT_FONT_SIZE = 16;
// Available type scale values
export const SCALE_VALUES = {
  1.067: "Minor Second",
  1.125: "Major Second",
  1.2: "Minor Third",
  1.25: "Major Third",
  1.333: "Perfect Fourth",
  1.414: "Augmented Fourth",
  1.5: "Perfect Fifth",
  1.618: "Golden Ratio"
};
// Minimum gap between the color picker icon and viewport
// edge required to place the color picker
export const MIN_GAP = 300;
// Notification snackbar visibility timeout
export const TIMEOUT = 10000;
