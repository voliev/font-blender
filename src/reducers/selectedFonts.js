import undoable, { includeAction } from "redux-undo";
import {
  ADD_SELECTED_FONT,
  CLEAR_SELECTED_FONTS,
  REMOVE_FONT_FROM_SELECTED,
  UNDO_DESELECTED
} from "../constants";
import { getFonts } from "./fonts";

const initialState = [];

export const selectedFonts = (state = initialState, action) => {
  switch (action.type) {
    case ADD_SELECTED_FONT:
      return [...state, action.family];
    case REMOVE_FONT_FROM_SELECTED:
      return state.filter(family => family !== action.family);
    case CLEAR_SELECTED_FONTS:
      return [];
    default:
      return state;
  }
};

export default undoable(selectedFonts, {
  limit: 1,
  filter: includeAction([REMOVE_FONT_FROM_SELECTED, CLEAR_SELECTED_FONTS]),
  undoType: UNDO_DESELECTED
});

export const getSelectedFontFamilies = state => state.selectedFonts.present;
export const getSelectedFonts = state => {
  const fonts = getFonts(state);

  return getSelectedFontFamilies(state).reduce((selected, family) => {
    selected[family] = fonts[family];
    return selected;
  }, {});
};
export const isSelected = (state, family) =>
  state.selectedFonts.present.includes(family);
