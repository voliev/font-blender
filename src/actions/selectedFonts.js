import {
  ADD_SELECTED_FONT,
  CLEAR_SELECTED_FONTS,
  REMOVE_FONT_FROM_SELECTED,
  UNDO_DESELECTED
} from "../constants";
import { startFetchFont } from "./fonts";
import { addNotification, removeNotification } from "./notifications";
import { getFonts } from "../reducers/fonts";

export const addFontToSelected = family => ({
  type: ADD_SELECTED_FONT,
  family
});

export const removeFontFromSelected = family => ({
  type: REMOVE_FONT_FROM_SELECTED,
  family
});

export const clearSelectedFonts = () => ({
  type: CLEAR_SELECTED_FONTS
});

export const startAddFontToSelected = family => {
  return (dispatch, getState) => {
    const state = getState();
    const fonts = getFonts(state);
    const font = fonts[family];

    if (font.fetched) {
      // Add font to selected if it has been already fetchedselected list
      dispatch(addFontToSelected(family));
    } else {
      // Fetch font first than add it to selected
      return dispatch(startFetchFont(family))
        .then(() => dispatch(addFontToSelected(family)))
        .catch(() => {});
    }
  };
};

export const undoDeselected = () => ({
  type: UNDO_DESELECTED
});

export const startRemoveFontFromSelected = family => {
  return dispatch => {
    dispatch(removeFontFromSelected(family));

    const text = `${family} font family has been removed`;
    const group = "selectedFonts";
    // Add notification to display it with "undo" option
    dispatch(addNotification(text, group));
  };
};

export const startClearSelectedFonts = () => {
  return (dispatch, getState) => {
    const state = getState();
    const selectedFonts = [...state.selectedFonts.present];

    dispatch(clearSelectedFonts());

    const text =
      selectedFonts.length > 1
        ? `${selectedFonts.length} font families have been removed`
        : `${selectedFonts.length} font family have been removed`;
    const group = "selectedFonts";
    // Add notification to display it with "undo" option
    dispatch(addNotification(text, group));
  };
};

export const startUndoDeselected = () => {
  return dispatch => {
    // Undo deselection
    dispatch(undoDeselected());

    const group = "selectedFonts";

    // Remove notification to hide "undo" option
    dispatch(removeNotification(group));
  };
};
