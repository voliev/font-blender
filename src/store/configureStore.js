import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import fonts from "../reducers/fonts";
import filters from "../reducers/filters";
import selectedFonts from "../reducers/selectedFonts";
import showcase from "../reducers/showcase";
import styles from "../reducers/styles";
import notifications from "../reducers/notifications";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export default () => {
  const store = createStore(
    combineReducers({
      fonts,
      filters,
      selectedFonts,
      showcase,
      styles,
      notifications
    }),
    composeEnhancers(applyMiddleware(thunk))
  );

  return store;
};
