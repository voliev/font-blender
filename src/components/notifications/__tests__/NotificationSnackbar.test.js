import React from "react";
import { mount, shallow } from "enzyme";
import configureStore from "redux-mock-store";
import NotificationSnackbarConnected, {
  NotificationSnackbar
} from "../NotificationSnackbar";
import UndoButton from "../UndoButton";
import selectedFonts from "../../../fixtures/selectedFonts";
import state from "../../../fixtures/state";
import { TIMEOUT } from "../../../constants";

const setup = (renderFn = shallow, propsOverrides = {}, connected = false) => {
  const props = {
    notification: {
      current: "Roboto font family has been removed",
      pending: ""
    },
    group: "selectedFonts",
    groupState:
      propsOverrides.group !== "clipboard" ? selectedFonts : undefined,
    removeNotification: jest.fn(),
    undoActionHandler:
      propsOverrides.group !== "clipboard" ? jest.fn() : undefined,
    ...propsOverrides
  };

  if (connected) {
    const mockStore = configureStore();
    const store = mockStore(state);

    return {
      ...props,
      wrapper: renderFn(
        <NotificationSnackbarConnected {...props} store={store} />
      )
    };
  }

  return {
    ...props,
    wrapper: renderFn(<NotificationSnackbar {...props} />)
  };
};

afterEach(() => {
  jest.clearAllMocks();
});

it("should render without crashing", () => {
  const { wrapper } = setup();

  expect(wrapper).toMatchSnapshot();
});

it("should receive corresponding props", () => {
  const {
    wrapper,
    notification,
    removeNotification,
    undoActionHandler,
    groupState
  } = setup(mount, undefined, true);
  const component = wrapper.find(NotificationSnackbar);

  expect(typeof component.prop("notification")).toBe("object");
  expect(component.prop("notification")).toEqual(notification);

  expect(typeof component.prop("group")).toBe("string");
  expect(component.prop("group")).toBe("selectedFonts");

  expect(typeof component.prop("removeNotification")).toBe("function");
  expect(component.prop("removeNotification")).toBe(removeNotification);

  expect(typeof component.prop("undoActionHandler")).toBe("function");
  expect(component.prop("undoActionHandler")).toBe(undoActionHandler);

  expect(typeof component.prop("groupState")).toBe("object");
  expect(component.prop("groupState").hasOwnProperty("length")).toBeTruthy();
  expect(component.prop("groupState")).toEqual(groupState);
});

it("should receive corresponding props", () => {
  const { wrapper, notification, removeNotification } = setup(
    mount,
    {
      notification: {
        current: "Copied to clipboard",
        pending: ""
      },
      group: "clipboard"
    },
    true
  );

  const component = wrapper.find(NotificationSnackbar);

  expect(typeof component.prop("notification")).toBe("object");
  expect(component.prop("notification")).toEqual(notification);

  expect(typeof component.prop("group")).toBe("string");
  expect(component.prop("group")).toBe("clipboard");

  expect(typeof component.prop("removeNotification")).toBe("function");
  expect(component.prop("removeNotification")).toBe(removeNotification);

  expect(component.prop("undoActionHandler")).toBeUndefined();

  expect(component.prop("groupState")).toBeUndefined();
});

it("should set corresponding state", () => {
  const { wrapper } = setup();

  expect(wrapper.state("stage")).toBe("elevate");
  expect(wrapper.state("timerId")).toBeUndefined();
});

it("should set corresponding component classes", () => {
  const { wrapper } = setup();

  expect(wrapper.hasClass("NotificationSnackbar")).toBeTruthy();
  expect(wrapper.hasClass("elevate")).toBeTruthy();
  expect(wrapper.hasClass("display")).toBeFalsy();
  expect(wrapper.hasClass("drop")).toBeFalsy();
  expect(wrapper.hasClass("undo")).toBeFalsy();

  wrapper.setState({
    stage: "display",
    timeId: 12345
  });

  expect(wrapper.hasClass("NotificationSnackbar")).toBeTruthy();
  expect(wrapper.hasClass("elevate")).toBeFalsy();
  expect(wrapper.hasClass("display")).toBeTruthy();
  expect(wrapper.hasClass("drop")).toBeFalsy();
  expect(wrapper.hasClass("undo")).toBeFalsy();

  wrapper.setState({
    stage: "drop",
    timeId: undefined
  });

  expect(wrapper.hasClass("NotificationSnackbar")).toBeTruthy();
  expect(wrapper.hasClass("elevate")).toBeFalsy();
  expect(wrapper.hasClass("display")).toBeFalsy();
  expect(wrapper.hasClass("drop")).toBeTruthy();
  expect(wrapper.hasClass("undo")).toBeFalsy();

  wrapper.setState({
    stage: "undo",
    timeId: undefined
  });

  expect(wrapper.hasClass("NotificationSnackbar")).toBeTruthy();
  expect(wrapper.hasClass("elevate")).toBeFalsy();
  expect(wrapper.hasClass("display")).toBeFalsy();
  expect(wrapper.hasClass("drop")).toBeFalsy();
  expect(wrapper.hasClass("undo")).toBeTruthy();
});

it("should render `undo` button", () => {
  const { wrapper, undoActionHandler } = setup(mount, undefined, true);

  expect(wrapper.find(NotificationSnackbar).prop("undoActionHandler")).toBe(
    undoActionHandler
  );
  expect(wrapper.find(UndoButton).length).toBe(1);
});

it("should NOT render `undo` button", () => {
  const { wrapper } = setup(
    mount,
    {
      notification: {
        current: "Copied to clipboard",
        pending: ""
      },
      group: "clipboard"
    },
    true
  );

  expect(wrapper.find(UndoButton).length).toBe(0);
});

it("should set a timer when a snackbar has been elevated", () => {
  const setTimer = jest.spyOn(NotificationSnackbar.prototype, "setTimer");
  const { wrapper } = setup();

  wrapper.instance().handleAnimationEnd();

  expect(setTimer).toHaveBeenCalledTimes(1);
  expect(wrapper.state("stage")).toBe("display");
  expect(wrapper.state("timerId")).toBeDefined();
});

it("should clear timer after display timeout", () => {
  jest.useFakeTimers();
  const setTimer = jest.spyOn(NotificationSnackbar.prototype, "setTimer");
  const clearTimer = jest.spyOn(NotificationSnackbar.prototype, "clearTimer");
  const { wrapper } = setup();

  expect(wrapper.state("stage")).toBe("elevate");
  expect(wrapper.state("timerId")).toBeUndefined();

  wrapper.instance().handleAnimationEnd();

  expect(setTimer).toHaveBeenCalledTimes(1);
  expect(wrapper.state("stage")).toBe("display");
  expect(wrapper.state("timerId")).toBeDefined();
  expect(clearTimer).toHaveBeenCalledTimes(0);

  jest.runTimersToTime(TIMEOUT);

  expect(clearTimer).toHaveBeenCalledTimes(1);
  expect(wrapper.state("stage")).toBe("drop");
  expect(wrapper.state("timerId")).toBeUndefined();
});

it("should remove notification on `drop` stage", () => {
  const { wrapper, removeNotification } = setup();

  wrapper.setState({ stage: "drop" });
  wrapper.instance().handleAnimationEnd();

  expect(removeNotification).toHaveBeenCalledTimes(1);
  expect(removeNotification).toHaveBeenCalledWith("selectedFonts");
});

it("should handle `undo` button click", () => {
  const { wrapper } = setup();

  wrapper.setState({
    stage: "display",
    timeId: 123
  });
  wrapper.instance().handleUndoButtonClick();

  expect(wrapper.state("stage")).toBe("undo");
  expect(wrapper.state("timerId")).toBeUndefined();
});

it("should call undoActionHandler on `undo` stage", () => {
  const { wrapper, undoActionHandler } = setup();

  wrapper.setState({ stage: "undo" });
  wrapper.instance().handleAnimationEnd();

  expect(undoActionHandler).toHaveBeenCalledTimes(1);
});

it("should handle `close` icon click", () => {
  const handleCloseIconClick = jest.spyOn(
    NotificationSnackbar.prototype,
    "handleCloseIconClick"
  );
  const clearTimer = jest.spyOn(NotificationSnackbar.prototype, "clearTimer");
  const { wrapper } = setup();

  wrapper.find(".close-icon").simulate("click");

  expect(handleCloseIconClick).toHaveBeenCalledTimes(1);
  expect(clearTimer).toHaveBeenCalledTimes(1);
  expect(clearTimer).toHaveBeenCalledWith("drop");
  expect(wrapper.state("stage")).toBe("drop");
  expect(wrapper.state("timerId")).toBeUndefined();
});

it("should clear timer on pending notification", () => {
  const { wrapper, notification } = setup();

  wrapper.setState({
    stage: "display",
    timerId: 123
  });

  expect(wrapper.state("timerId")).toBe(123);

  wrapper.setProps({
    groupState: selectedFonts.slice(1),
    notification: {
      ...notification,
      pending: "Lato font family has been removed"
    }
  });

  expect(wrapper.state("timerId")).toBeUndefined();
});

it("should clear timer on group state change", () => {
  const clearTimer = jest.spyOn(NotificationSnackbar.prototype, "clearTimer");
  const { wrapper } = setup();

  wrapper.setState({
    stage: "display",
    timerId: 123
  });

  expect(wrapper.state("timerId")).toBe(123);

  wrapper.setProps({
    groupState: [...selectedFonts, "Roboto"]
  });

  expect(clearTimer).toHaveBeenCalledTimes(1);
  expect(clearTimer).toHaveBeenCalledWith("drop");
  expect(wrapper.state("timerId")).toBeUndefined();
});

it("should NOT clear timer and drop notification", () => {
  const clearTimer = jest.spyOn(NotificationSnackbar.prototype, "clearTimer");
  const { wrapper } = setup(shallow, {
    notification: {
      current: "Copied to clipboard",
      pending: ""
    },
    group: "clipboard"
  });

  wrapper.instance().componentDidUpdate();

  expect(clearTimer).not.toHaveBeenCalled();
});
