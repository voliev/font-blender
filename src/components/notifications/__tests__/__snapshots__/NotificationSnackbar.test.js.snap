// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<div
  className="NotificationSnackbar elevate"
  onAnimationEnd={[Function]}
>
  <p
    className="notification"
  >
    Roboto font family has been removed
  </p>
  <div
    className="controls"
  >
    <UndoButton
      onClickHandler={[Function]}
    />
    <button
      className="btn btn-round-small close-icon"
      onClick={[Function]}
      title="Close"
      value="close"
    >
      <ClearOutlinedIcon
        fontSize="small"
      />
    </button>
  </div>
</div>
`;
